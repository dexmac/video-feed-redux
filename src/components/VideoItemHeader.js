import React from "react";
import VideoItemTitle from "./VideoItemTitle";
import VideoItemViewsNumber from "./VideoItemViewsNumber";
import PropTypes from "prop-types";

const VideoItemHeader = ({ title = "", viewsNumber }) => (
  <div className="video-item-header">
    <VideoItemTitle className="video-item-title" title={title} />
    <VideoItemViewsNumber
      className="video-item-views-number"
      viewsNumber={viewsNumber}
    />
  </div>
);

VideoItemHeader.propTypes = {
  title: PropTypes.string,
  viewsNumber: PropTypes.number
};

export default VideoItemHeader;
