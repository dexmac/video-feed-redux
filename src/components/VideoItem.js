import React from "react";
import PropTypes from "prop-types";
import ReactPlayer from "react-player";
import VideoItemHeader from "./VideoItemHeader";

const VideoItem = ({
  item = {},
  onVideoLoaded = () => {},
  onVideoError = () => {}
}) => (
  <div className="video-item">
    <VideoItemHeader
      title={item.title ? item.title : ""}
      viewsNumber={item.views}
    />
    <div className="player-wrapper">
      <ReactPlayer
        className="react-player"
        url={item.url ? item.url : null}
        key={item.videoId ? item.videoId : null}
        width="100%"
        height="100%"
        loop={true}
        onError={e => {
          onVideoError(item.url ? e + " for url: " + item.url : e, item);
        }}
        onReady={() => {
          onVideoLoaded(item);
        }}
        playing
        muted
        playsinline
        config={{
          file: {
            attributes: { muted: "true", autoPlay: "true" }
          }
        }}
      />
    </div>
  </div>
);

VideoItem.propTypes = {
  onVideoLoaded: PropTypes.func,
  onVideoError: PropTypes.func,
  item: PropTypes.shape({
    url: PropTypes.string,
    source: PropTypes.string,
    videoId: PropTypes.string,
    title: PropTypes.string,
    views: PropTypes.number
  })
};

export default VideoItem;
