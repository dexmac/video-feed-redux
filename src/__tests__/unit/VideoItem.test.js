import React from "react";
import jest from "jest-mock";
import VideoItem from "../../components/VideoItem";
import VideoItemHeader from "../../components/VideoItemHeader";
import ReactPlayer from "react-player";

describe("VideoItem", () => {
  let props;
  let mountedVideoItem;

  /**
   * Clearing props before each test
   */
  beforeEach(() => {
    props = {
      item: {}
    };
    mountedVideoItem = undefined;
  });

  /**
   * Shallow rendering vs. snapshot test
   */
  it("renders correctly && matches snapshot", () => {
    const tree = shallow(<VideoItem />);
    expect(tree).toMatchSnapshot();
  });

  /**
   * Mounts the component using Enzyme if it is not mounted yet
   * @returns {object} - returns the mounted component
   */
  const videoItem = () => {
    if (!mountedVideoItem) {
      mountedVideoItem = mount(<VideoItem {...props} />);
    }
    return mountedVideoItem;
  };

  describe("when `title` is defined", () => {
    beforeEach(() => {
      props.item.title = "myTitle";
    });

    it("sets the rendered text to the same value as `title`", () => {
      const header = videoItem().find(VideoItemHeader);
      expect(header.props().title).toBe(props.item.title);
    });
  });

  describe("when `title` is undefined", () => {
    beforeEach(() => {
      props.item.title = undefined;
    });

    it("sets the rendered text to empty", () => {
      const header = videoItem().find(VideoItemHeader);
      expect(header.props().title).toBe("");
    });
  });

  describe("when `views` is defined", () => {
    beforeEach(() => {
      props.item.views = 12345;
    });

    it("sets the rendered text to the same value as `views`", () => {
      const header = videoItem().find(VideoItemHeader);
      expect(header.props().viewsNumber).toBe(props.item.views);
    });
  });

  describe("when `views` is undefined", () => {
    beforeEach(() => {
      props.item.views = undefined;
    });

    it("sets the rendered text to empty", () => {
      const header = videoItem().find(VideoItemHeader);
      expect(header.props().viewsNumber).toBe(undefined);
    });
  });

  describe("when `url` is defined", () => {
    beforeEach(() => {
      props.item.url = "myURL";
    });

    it("sets the rendered text to the same value as `url`", () => {
      const player = videoItem().find(ReactPlayer);
      expect(player.props().url).toBe(props.item.url);
    });
  });

  describe("when `url` is undefined", () => {
    beforeEach(() => {
      props.item.url = undefined;
    });

    it("sets the rendered text to be null", () => {
      const player = videoItem().find(ReactPlayer);
      expect(player.props().url).toBe(null);
    });
  });

  describe("when an item is rendered", () => {
    beforeEach(() => {
      props.item.url = undefined;
      props.item.videoId = "test";
    });

    it("always renders wrapper div around video player", () => {
      const wrapperDiv = videoItem().find("div.player-wrapper");
      expect(wrapperDiv.length).toBe(1);
    });

    it("always renders a react player first", () => {
      const player = videoItem().find(ReactPlayer);
      expect(player.length).toBe(1);
    });

    it("Should run onload and onerror methods without crashing", () => {
      const player = videoItem()
        .find(ReactPlayer)
        .instance().props;
      player.onReady();
      player.onError();
      props.item.url = "test";
      player.onError();
    });
  });
});
