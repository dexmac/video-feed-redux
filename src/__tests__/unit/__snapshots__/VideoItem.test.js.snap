// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoItem renders correctly && matches snapshot 1`] = `
<div
  className="video-item"
>
  <VideoItemHeader
    title=""
  />
  <div
    className="player-wrapper"
  >
    <ReactPlayer
      className="react-player"
      config={
        Object {
          "file": Object {
            "attributes": Object {
              "autoPlay": "true",
              "muted": "true",
            },
          },
        }
      }
      controls={false}
      height="100%"
      key="null"
      loop={true}
      muted={true}
      onBuffer={[Function]}
      onDuration={[Function]}
      onEnded={[Function]}
      onError={[Function]}
      onPause={[Function]}
      onPlay={[Function]}
      onProgress={[Function]}
      onReady={[Function]}
      onSeek={[Function]}
      onStart={[Function]}
      playbackRate={1}
      playing={true}
      playsinline={true}
      progressInterval={1000}
      style={Object {}}
      url={null}
      volume={null}
      width="100%"
      wrapper="div"
    />
  </div>
</div>
`;
